File: quotebot-python/quotebot/ad_space.py
from quotebot.tech_blogs import TechBlogs


class AdSpace:

    _cache = {}

    @staticmethod
    def get_ad_spaces():
        if "blogs list" in AdSpace._cache:
            return AdSpace._cache["blogs list"]
        # FIXME : only return blogs that start with a 'T'
        list_all_blogs = TechBlogs.list_all_blogs()
        AdSpace._cache["blogs list"] = list_all_blogs
        return list_all_blogs

File: quotebot-elixir/lib/ad_space.ex
defmodule Quotebot.AdSpace do
  @moduledoc """
  AdSpace module with static cache.
  Stateful behavior: first call different from next calls.
  """

  use Agent

  def start_link(_opts) do
    Agent.start_link(fn -> [] end, name: __MODULE__)
  end

  def get_ad_spaces do
    # Try to use Agent if started, otherwise use process dictionary
    cache =
      case Process.whereis(__MODULE__) do
        nil -> Process.get(:ad_space_cache, [])
        _pid -> Agent.get(__MODULE__, & &1)
      end

    if cache != [] do
      cache
    else
      # FIXME: only return blogs that start with a 'T'
      blogs = Quotebot.TechBlogs.list_all_blogs()

      # Store in cache
      case Process.whereis(__MODULE__) do
        nil -> Process.put(:ad_space_cache, blogs)
        _pid -> Agent.update(__MODULE__, fn _ -> blogs end)
      end

      blogs
    end
  end

  def clear_cache do
    case Process.whereis(__MODULE__) do
      nil -> Process.put(:ad_space_cache, [])
      _pid -> Agent.update(__MODULE__, fn _ -> [] end)
    end
  end
end
